### PR TITLE
FullofSkittles Custom pAI & Silicon Fluff Tweaks

### PR DESCRIPTION
**Pull request by FalseIncarnate**
Adds in FullofSkittles's custom pAI for LUNE.
### Standing:
image
### Resting:
image
### On-head / In-hand sprite:
image

Adds support for custom pAI sprites.

Tweaks code for custom cyborg, AI, and hologram sprites to be more standardized, which made adding pAI support that much easier.
- Cyborg, AI, hologram, and pAI custom sprites also are no longer restricted to a specific real_name. As long as your ckey matches, you can equip your custom borg/AI/hologram/pAI sprite on any borg/AI/pAI.

Adds support for custom maint drone sprites. All drones have a one-use verb which will attempt to equip a custom sprite if their ckey has one.
- It is up to the head admins if we want to let people actually get custom maint drone sprites, but we at least have the systems for it now.

As a result of the changes, the custom_sprites.txt file will need to be updated to the new standardized format (been working with Alffd on that anyways so he's aware).
- I have taken the liberty of updating the example custom_sprites.txt to explain and match this new format for our downstreams and other coders.

:cl:
rscadd: Adds FullofSkittles's custom pAI sprites. Thanks FoS!
tweak: Custom sprites for cyborgs, AI, AI holograms, and pAI units no longer are restricted to a single name. Get your money's worth by using that borg sprite regardless of your version number!
/:cl:

### CODE REVIEW

**Comment by Jovaniph** [comment]
Cute

**Comment by taukausanake** [comment]
Amazing! Also, those resting icons. Those are really weird. Let alone the vomiting of what I'm sure is liquid skittles. Shouldn't there be some spriters rule for directional icons actually be directional?

**Comment by FalseIncarnate** [comment]
She did the resting ones like she did because otherwise she'd be asking for special emotes that only work for her pAI too. I mentioned this to her when she first sent them to me, but compared to the alternative I think this is reasonable.

**Comment by Fullofskittles** [comment]
> Amazing! Also, those resting icons. Those are really weird. Let alone the vomiting of what I'm sure is liquid skittles. Shouldn't there be some spriters rule for directional icons actually be directional?

Most mobs that have resting sprites have a single sprite instead of "proper" directional sprites. I think there are only four pAI sprites that actually have resting sprites that aren't just a single static pose. Why would there be a rule for resting sprites? I cant really think of any reason that would require a rule like that.

**Comment by Spacemanspark** [comment]
...hrmmmmmm
I change my mind again.
I want a pAI fluff item now.
Also, that's a really kickass sprite.

**Comment by tigercat2000** [comment]
@Fullofskittles Generally, different "actions" or states of being shouldn't be just stuffed together into one "icon" as different directions

It makes it harder to actually tell what's inside an icon and what you'll get
it's actually a massive problem with turfs

They should have seperate icon_states

**Comment by taukausanake** [comment]
Also, where I do like the icons for head/in-hands being wings on your back, shouldn't the icons match what you'd think it would? Cool wings still but I think what Tiger said about turfs also counts for this

**Comment by FalseIncarnate** [comment]
She didn't have in-hands and I didn't want to have invisible in-hands for this. All other pAI forms have in-hand and on-head sprites, so making this the exception didn't feel right.

**Comment by Draconic-saint** [comment]
it fucking pukes rainbows 💯 times 👍

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 1]
```diff
@@ -299,22 +299,59 @@ var/list/ai_verbs_default = list(
...
 
-			if(Entry.len < 2)
-				continue
-
-			if(Entry.len < 3 && Entry[1] == ckey && Entry[2] == real_name)
-				custom_sprite = 1 //They're in the list? Custom sprite time
-				icon = 'icons/mob/custom-synthetic.dmi'
-
+			if(Entry.len < 2 || Entry[1] != "ai")			//ignore incorrectly formatted entries or entries that aren't marked for AI
+				continue;
```
Comment (Fox-McCloud) [inline code/modules/mob/living/silicon/ai/ai.dm]: `continue;`

File: code/modules/mob/living/silicon/pai/pai.dm [hunk 2]
```diff
@@ -394,17 +395,47 @@
...
+		var/lines = splittext(file, "\n")
+
+		for(var/line in lines)
+		// split & clean up
+			var/list/Entry = splittext(line, ":")
+			for(var/i = 1 to Entry.len)
+				Entry[i] = trim(Entry[i])
+
+			if(Entry.len < 2 || Entry[1] != "pai")			//ignore incorrectly formatted entries or entries that aren't marked for pAI
+				continue;
```
Comment (Fox-McCloud) [inline code/modules/mob/living/silicon/pai/pai.dm]: `continue`

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 3]
```diff
@@ -180,16 +180,15 @@ var/list/robot_verbs_default = list(
...
 			for(var/line in lines)
 			// split & clean up
-				var/list/Entry = splittext(line, ";")
+				var/list/Entry = splittext(line, ":")
 				for(var/i = 1 to Entry.len)
 					Entry[i] = trim(Entry[i])
 
-				if(Entry.len < 2)
+				if(Entry.len < 2 || Entry[1] != "cyborg")		//ignore incorrectly formatted entries or entries that aren't marked for cyborg
 					continue;
```
Comment (Fox-McCloud) [inline code/modules/mob/living/silicon/robot/robot.dm]: Reeee

**Comment by TravellingMerchant** [comment]
I am all for this. That is a lovely Sprite!

File: code/modules/mob/living/silicon/pai/pai.dm [hunk 4]
```diff
@@ -394,17 +395,47 @@
...
+			var/list/Entry = splittext(line, ":")
+			for(var/i = 1 to Entry.len)
+				Entry[i] = trim(Entry[i])
+
+			if(Entry.len < 2 || Entry[1] != "pai")			//ignore incorrectly formatted entries or entries that aren't marked for pAI
+				continue
+
+			if(Entry[2] == ckey)							//They're in the list? Custom sprite time, var and icon change required
+				custom_sprite = 1
+				my_choices["Custom"] = "[ckey]-pai"
```
Comment (Crazylemon64) [inline code/modules/mob/living/silicon/pai/pai.dm]: mildly redundant but whatever

Comment (FalseIncarnate) [inline code/modules/mob/living/silicon/pai/pai.dm]: So it is... must have been leftover from when I was initially thinking of adding support for multiple custom sprite sets on a single ckey (which I decided not to do in this PR) that I missed.

File: code/modules/mob/living/silicon/robot/drone/drone_abilities.dm [hunk 5]
```diff
@@ -47,4 +47,55 @@
...
+			if (Entry[2] == ckey) //Custom holograms
+				custom_sprite = 1  // option is given in hologram menu
+
+	if(!custom_sprite)
+		to_chat(src, "<span class='warning'>Error 404: Custom chassis not found. Revoking customization option.</span>")
+	else
+		icon = 'icons/mob/custom_synthetic/custom-synthetic.dmi'
+		icon_state = "[ckey]-drone"
+		to_chat(src, "<span class='notice'>You reconfigure your chassis and improve the station through your new aesthetics.</span>")
+	verbs -= /mob/living/silicon/robot/drone/verb/customize
```
Comment (Crazylemon64) [inline code/modules/mob/living/silicon/robot/drone/drone_abilities.dm]: If you wanted things smoother you could put this in the drone's `login` but it's not really important

**Pull request merged**